### DISCUSSION TITLE
Question on printing variables to the console

### DISCUSSION THREAD

**Discussion by tmob03** [ArgosOfIthica/scalu#26] [Q&A] [answered]
In this post you made https://www.reddit.com/r/Tf2Scripts/comments/im624x/scalu_a_programming_language_for_configuration/
you mention scalu has the ability to print variable values to the console, is this just simple if (x == 6) { [echo 6] } or is there something to not have to pre-define all the different values and just do [echo X] or something?

**Comment by ArgosOfIthaca** [comment, marked as answer]
Good catch, I don't mention this very well in the documentation.

There's 2 methods. There's the debug operator `?`, which when used like:

`a = ?a`

Will print the value of `a` to the console, but in binary.

I've been too busy to write a standard library, where the 2nd alternative would have gone, but you can quite simply implement `print` yourself.

```
service print {
    jump (a) {{[echo 0]} {[echo 1]} {[echo 2]} .... }
}
```
While verbose, this is very fast, as jump determines which block to execute by scanning the bits of `a`; it can be as fast as a single comparison. Then, I'd generally recommend setting the value of `a` with whatever value you want to examine rather than creating this service for every variable.

**Reply by tmob03** [reply]
Thanks!

**Reply by tmob03** [reply]
@ArgosOfIthica I'd probably be willing to create a standard library, what all do you think would be useful in it?

**Reply by tmob03** [reply]
With no function argument and function return equivalent I can’t imagine there’s a whole lot you could do in a library, just things like printint(0-255)and printchar as seen in hangman, unless I’m really underestimating what scalu currently can do…

**Reply by ArgosOfIthaca** [reply]
> With no function argument and function return equivalent I can’t imagine there’s a whole lot you could do in a library, just things like printint(0-255)and printchar as seen in hangman, unless I’m really underestimating what scalu currently can do…

 That's true, scalu does not implement a built-in stack, so true functions aren't feasible at the language level. However, that has very little implication for writing useful services; many assembly languages also lack any built-in notion of a stack. As an example, services I'd consider implementing might be multiplication, variable bitshifts, xor prng, modulus, bigints, etc (That's not even mentioning any engine or game-specific utilities.). You can simply declare specific input and output variables that anyone writing code against your library will already know about.
